### PR TITLE
Calculate interactions between feature groups

### PR DESCRIPTION
**Pull request by MaxHalford**
This PR addresses a very good suggestion from @oegedijk in #776.

I'm a bit tired so not 100% sure inheriting from `TransformerUnion` is a good idea as is, but it works. Maybe there's some edge cases I'm not seeing but we can catch those later.

### CODE REVIEW

**Comment by oegedijk** [comment]
Looks great! Very nice clean implementation. 

I would probably add a deprecation error to the old `FeatureExtender` and keep it around for a while though...

File: river/compose/product.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+from . import union
+
+__all__ = ["TransformerProduct"]
+
+
+class TransformerProduct(union.TransformerUnion):
+    """Computes interactions between the outputs of a set transformers.
+
+    This is for when you want to add interaction terms between groups of features. It may also
+    be used an alternative to `feature_extraction.PolynomialExtender` when the latter is overkill.
```
Comment (oegedijk) [inline river/compose/product.py:14]: should be changed to PolynomialFeatures

**Comment by oegedijk** [comment]
I was also thinking whether there is a way to make it even more readable with R-style function definitions (i.e. basically automatically insert all the `Select()`). 

So you could do something like `Formula("a + b*x")` which would then generate `Select("a") + Select("b") * Select("x")` under the hood.

File: river/compose/product.py (hunk 1 above)
Comment (MaxHalford) [reply river/compose/product.py:14]: Actually I'm keeping the name `PolynomialExtender`

**Comment by MaxHalford** [comment]
Mmm well I would say we can keep `PolynomialExtender` as is. It's still a nice thing to have around.

**Comment by MaxHalford** [comment]
> So you could do something like Formula("a + b*x") which would then generate Select("a") + Select("b") * Select("x") under the hood.

Indeed! That's what Patsy does in the Python ecosystem. I would handle this in a separate PR though. It requires some string parsing. I'm not 100% on how to get it done properly. But at least now we have some good building blocks to build it.

**Comment by oegedijk** [comment]
> Indeed! That's what Patsy does in the Python ecosystem. I would handle this in a separate PR though. It requires some string parsing. I'm not 100% on how to get it done properly. But at least now we have some good building blocks to build it.

Yeah definitely out of scope for this PR, would be a great readability definition though!

**Comment by oegedijk** [comment]
> Mmm well I would say we can keep `PolynomialExtender` as is. It's still a nice thing to have around.

Ah, hadn't seen the revert renaming commit. I do think `PolynomialFeatures` is a more descriptive name, but also better not to have too many unnecessary breaking changes :)

**Comment by oegedijk** [comment]
Stress testing a bit, this seems to work:

```
group_1 = Select('a', 'b') | OneHotEncoder(sparse=True)
group_2 = Select('x', 'y')
x = dict(a=1, b=2, x=3, y=4, z=5)


tp1 = TransformerUnion(group_1, group_2,TransformerProduct(group_1, group_2))
tp1.transform_one(x)
```

But this does not:

```
tp2 = group_1 + group_2 + group_1 * group_2
tp2.transform_one(dict(a=1, b=2, x=4, y=4, z=5))
```

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_6101/2982673641.py in <module>
      1 tp2 = group_1 + group_2 + group_1 * group_2
----> 2 tp2.transform_one(dict(a=1, b=2, x=4, y=4, z=5))

/opt/conda/lib/python3.7/site-packages/river/compose/union.py in transform_one(self, x)
    231         return dict(
    232             collections.ChainMap(
--> 233                 *(t.transform_one(x) for t in self.transformers.values())
    234             )
    235         )

/opt/conda/lib/python3.7/site-packages/river/compose/union.py in <genexpr>(.0)
    231         return dict(
    232             collections.ChainMap(
--> 233                 *(t.transform_one(x) for t in self.transformers.values())
    234             )
    235         )

/opt/conda/lib/python3.7/site-packages/river/compose/grouper.py in transform_one(self, x)
     47 
     48     def transform_one(self, x):
---> 49         key = self._get_key(x)
     50         return self.transformers[key].transform_one(x)

/opt/conda/lib/python3.7/site-packages/river/compose/grouper.py in _get_key(self, x)
     39 
     40     def _get_key(self, x):
---> 41         return "_".join(str(x[k]) for k in self.by)
     42 
     43     def learn_one(self, x):

/opt/conda/lib/python3.7/site-packages/river/compose/grouper.py in <genexpr>(.0)
     39 
     40     def _get_key(self, x):
---> 41         return "_".join(str(x[k]) for k in self.by)
     42 
     43     def learn_one(self, x):

KeyError: Pipeline (
  Select (
    a
    b
  ),
  OneHotEncoder (
    sparse=True
  )
)
```

**Comment by MaxHalford** [comment]
Good catch! Fixing.

**Comment by oegedijk** [comment]
Looks good!

**Pull request merged**